### PR TITLE
cli: Add unsafe-remove-dead-replicas command

### PR DESCRIPTION
**Pull request by bdarnell**
This command can be used to (inconsistently) salvage a single replica
from a cluster that has lost multiple nodes.

Release note: None

### CODE REVIEW

**Review by a-robinson — DISMISSED** [review]
:lgtm: but do we not also have to update the corresponding meta range entries?

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28111#-:-LIm2lEk-bDUA3U5TlF5:b-dq940o)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cli/debug.go, line 1190 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28111#-LIm1iXe8VvZRN9cq4yl:-LIm1iXe8VvZRN9cq4ym:bxg03it) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01f4d0c1271f6587c6efc472100a7bd0601a235/pkg/cli/debug.go#L1190)):*
> ```Go
> 	ctx := context.Background()
> 
> 	if len(args) != 1 {
> ```

You should be able to use the `Args: cobra.ExactArgs(1)` field on the `cobra.Command` to achieve this in a more consistent way.
___
*[pkg/cli/debug.go, line 1212 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28111#-LIm25wy0BIO33d32wgI:-LIm25wy0BIO33d32wgJ:bxq0d6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01f4d0c1271f6587c6efc472100a7bd0601a235/pkg/cli/debug.go#L1212)):*
> ```Go
> 
> 	if _, ok := deadStoreIDs[storeIdent.StoreID]; ok {
> 		return errors.New("This store's ID marked as dead, aborting")
> ```

Very minor, but if anyone hits this it'd probably be helpful for the error to include the store ID.

**Comment by bdarnell** [comment]
The meta entries will take care of themselves. We're already fairly tolerant of stale meta entries. And as soon as one of these single-replica ranges finds itself in a multi-node cluster, it will be upreplicated and the meta entries will be overwritten.

**Comment by tschottdorf** [comment]
Wouldn't this feel better if we had an acceptance test for this? I understand that you wrote this for an acute need, but it's quite scary to try this on anything when there's zero CI, and once it's committed we'll want to reuse it in similar situations.

**Comment by bdarnell** [comment]
Yeah. I've tested it manually but we should definitely build out some sort of test once the acute need has passed.

**Comment by tschottdorf** [comment]
From my experience, merging something without a test means there won't ever be a test. If it's critical for your progress to get this merged then go ahead, but otherwise I'd prefer to leave this sit unless a test can be added.

**Comment by bdarnell** [comment]
OK, added a test and also a new commit that handles intents which might have been left behind on the range descriptor (which happened on the cluster that motivated this work).

File: pkg/cli/debug.go [hunk 1]
```diff
@@ -1143,6 +1144,173 @@ func runDebugSyncTest(cmd *cobra.Command, args []string) error {
...
+		return err
+	} else if batch == nil {
+		fmt.Printf("Nothing to do\n")
+		return nil
+	}
+	defer batch.Close()
+
+	fmt.Printf("Proceed with the above rewrites? [y/N] ")
+
+	reader := bufio.NewReader(os.Stdin)
```
Comment (tschottdorf) [inline pkg/cli/debug.go]: Immaterial, but

```go
var input string
if _, err := fmt.Scanln(&input); err != nil || strings.ToLower(input) != "y" {
  os.Exit(1)    
} 
 ```

File: pkg/cli/debug.go [hunk 2]
```diff
@@ -1143,6 +1144,173 @@ func runDebugSyncTest(cmd *cobra.Command, args []string) error {
...
+		return nil, err
+	}
+
+	if len(newDescs) == 0 {
+		return nil, nil
+	}
+
+	batch := db.NewBatch()
+	for _, desc := range newDescs {
+		var ms enginepb.MVCCStats
```
Comment (tschottdorf) [inline pkg/cli/debug.go]: If you're not going to use `ms` at the end, pass `nil`.

File: pkg/cli/debug_test.go [hunk 3]
```diff
@@ -119,3 +125,115 @@ func TestOpenReadOnlyStore(t *testing.T) {
...
+	}()
+
+	// Now that the data is salvaged, we can restart the cluster. The
+	// nodes with the in-memory stores will be assigned new node IDs 4
+	// and 5. StartTestCluster will even wait for all the ranges to be
+	// replicated to the new nodes.
+	tc := testcluster.StartTestCluster(t, 3, clusterArgs)
+	defer tc.Stopper().Stop(ctx)
+	s := sqlutils.MakeSQLRunner(tc.Conns[0])
+	row := s.QueryRow(t, "select replicas from [show experimental_ranges from table system.namespace] limit 1")
```
Comment (tschottdorf) [inline pkg/cli/debug_test.go]: Can you create a table before the crash and read it back here?

**Review by tbg — APPROVED** [review]
LGTM

**Review by bdarnell — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28111)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cli/debug.go, line 1190 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28111#-LIm1iXe8VvZRN9cq4yl:-LIvsRzZ3JesRrVZy14R:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01f4d0c1271f6587c6efc472100a7bd0601a235/pkg/cli/debug.go#L1190)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

You should be able to use the `Args: cobra.ExactArgs(1)` field on the `cobra.Command` to achieve this in a more consistent way.
</blockquote></details>

Done.
___
*[pkg/cli/debug.go, line 1212 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28111#-LIm25wy0BIO33d32wgI:-LIvsU2S6F7QrTX65ovR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01f4d0c1271f6587c6efc472100a7bd0601a235/pkg/cli/debug.go#L1212)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Very minor, but if anyone hits this it'd probably be helpful for the error to include the store ID.
</blockquote></details>

Done.
___
*[pkg/cli/debug.go, line 1208 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/28111#-LIlsh5gfTIIjf92sydF-r3-1208:-LIvsSZi6DFWwuwZzibH:bhsbpkq) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf9e0b42fcb471f4b00519cd435eb580adb4a6a2/pkg/cli/debug.go#L1208)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Immaterial, but

```go
var input string
if _, err := fmt.Scanln(&input); err != nil || strings.ToLower(input) != "y" {
  os.Exit(1)    
} 
 ```
</blockquote></details>

Leaving this as-is.
___
*[pkg/cli/debug.go, line 1281 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/28111#-LIlsh5gfTIIjf92sydF-r3-1281:-LIvsU_oED03LEM9ooxC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf9e0b42fcb471f4b00519cd435eb580adb4a6a2/pkg/cli/debug.go#L1281)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

If you're not going to use `ms` at the end, pass `nil`.
</blockquote></details>

Done.
___
*[pkg/cli/debug_test.go, line 233 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/28111#-LIqP9Aj8BSumLJfC1kc-r3-233:-LIvsaCyC4aZeCNOWt4v:bdgs15f) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf9e0b42fcb471f4b00519cd435eb580adb4a6a2/pkg/cli/debug_test.go#L233)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Can you create a table before the crash and read it back here?
</blockquote></details>

Creating a table triggers a "too many ranges at startup" error in TestCluster, but I did a settings write so we can make sure that writes from before the crash can get persisted.

**Comment by bdarnell** [comment]
bors r=tschottdorf

**Comment by bdarnell** [comment]
Lint failure in my last update because I didn't wait for the last CI run.

**Comment by bdarnell** [comment]
The new test has a rare timing-related failure (which appears to show up only under `stressrace`, but does so immediately there). If the lease was on one of the dead nodes, this assertion can fire on restart:

https://github.com/cockroachdb/cockroach/blob/8044dea9c5c210c3ddb238a7edd446a93013e2e2/pkg/storage/replica.go#L1394

Interestingly, this didn't happen on the production cluster where this tool was used. I think in real-world situations the leaseholder's liveness would have expired, so we'd bypass this path. 

I think instead of tweaking things to avoid hitting the assertion (for example, by rewriting the lease when we rewrite the range descriptor), we should just downgrade it to a log.Error. (The assertion is already bypassed for expired leases, so if we do care about this as an assertion it should be happening at a higher level).

**Pull request merged**